tools/js-optimizer-babel.js

```diff
@@ -0,0 +1,130 @@
+// vim: set ts=2 et sw=2:
+
+/* 
+
+A version of the JS Optimizer that uses Babel
+
+*/
+
+var babel = require("babel-core");
+
+// *** Environment setup code ***
+
+var ENVIRONMENT_IS_NODE = typeof process === 'object';
+var ENVIRONMENT_IS_WEB = typeof window === 'object';
+var ENVIRONMENT_IS_WORKER = typeof importScripts === 'function';
+var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER;
+
+if (ENVIRONMENT_IS_NODE) {
+  // Expose functionality in the same simple way that the shells work
+  // Note that we pollute the global namespace here, otherwise we break in node
+  print = function(x) {
+    process['stdout'].write(x + '\n');
+  };
+  printErr = function(x) {
+    process['stderr'].write(x + '\n');
+  };
+
+  var nodeFS = require('fs');
+  var nodePath = require('path');
+
+  if (!nodeFS.existsSync) {
+    nodeFS.existsSync = function(path) {
+      try {
+        return !!nodeFS.readFileSync(path);
+      } catch(e) {
+        return false;
+      }
+    }
+  }
+
+  function find(filename) {
+    var prefixes = [nodePath.join(__dirname, '..', 'src'), process.cwd()];
+    for (var i = 0; i < prefixes.length; ++i) {
+      var combined = nodePath.join(prefixes[i], filename);
+      if (nodeFS.existsSync(combined)) {
+        return combined;
+      }
+    }
+    return filename;
+  }
+
+  read = function(filename) {
+    var absolute = find(filename);
+    return nodeFS['readFileSync'](absolute).toString();
+  };
+
+  load = function(f) {
+    globalEval(read(f));
+  };
+
+  arguments_ = process['argv'].slice(2);
+
+} else if (ENVIRONMENT_IS_SHELL) {
+  // Polyfill over SpiderMonkey/V8 differences
+  if (!this['read']) {
+    this['read'] = function(f) { snarf(f) };
+  }
+
+  if (typeof scriptArgs != 'undefined') {
+    arguments_ = scriptArgs;
+  } else if (typeof arguments != 'undefined') {
+    arguments_ = arguments;
+  }
+
+} else if (ENVIRONMENT_IS_WEB) {
+  this['print'] = printErr = function(x) {
+    console.log(x);
+  };
+
+  this['read'] = function(url) {
+    var xhr = new XMLHttpRequest();
+    xhr.open('GET', url, false);
+    xhr.send(null);
+    return xhr.responseText;
+  };
+
+  if (this['arguments']) {
+    arguments_ = arguments;
+  }
+} else if (ENVIRONMENT_IS_WORKER) {
+  // We can do very little here...
+
+  this['load'] = importScripts;
+
+} else {
+  throw 'Unknown runtime environment. Where are we?';
+}
+
+// *** MAIN ***
+
+// Process the incoming arguments
+
+var arguments_ = process['argv'].slice(2);
+
+arguments_ = arguments_.filter(function (arg) {
+  if (!/^--/.test(arg)) return true;
+  
+  if (arg === '--debug') debug = true;
+  else throw new Error('Unrecognized flag: ' + arg);
+});
+
+var src = read(arguments_[0]);
+
+function hasArg( arg )
+{
+  return arguments_.indexOf( arg ) > -1;
+}
+
+// Run through Babel
+
+var options = {
+  comments: !hasArg('minifyWhitespace'),
+  minified: hasArg('minifyWhitespace'),
```

Comment:
reading the babel docs, I think we also want `compact` to be true when we want to minify whitespace

Reply:
And if this is fast enough, maybe we should do it too (at least in higher opt levels) https://babeljs.io/docs/plugins/minify-mangle-names/

Reply:
You may be right. I had thought `compact` was implied by the `minified` option, but I probably mixed that up with another Babel option.

I guess going forward, we will have to decide how much we want to do with each of our passes. There's no overall minimise-everything pass that I could see (I guess the closure option usually does that instead?) But I haven't been using Closure because it's too aggressive.

I personally would prefer a Babel pass for maximal minification. minify-mangle-names shouldn't go in the minifyWhitespace pass, but instead the minifyGlobals and minifyLocals passes?

Reply:
Hmm, good points. I think for now we should just do `compact` to minify whitespace, and any other optimization we need to think more about.

Reply:
(and can leave to later)

Reply:
Setting `minified` automatically sets `compact`.